Fix turning, interacting with and shifting vehicles on ramps

joveeater opened the pull request:
#### Summary

SUMMARY: [Bugfixes] "Allow turning on ramps, interacting with vehicles across z-levels and prevent disappearing parts"

#### Purpose of change

Fixes #1212 by reallowing turning on ramps. It also fixes two other problems mentioned in that issue. Attempting to interact with vehicle parts on a different z level to the player won't cause an error and vehicles on ramps will no longer half disappear when loading or shifting submaps.

#### Describe the solution

Currently the code checks every part to see if it's on a ramp and move it up or down accordingly. Turning could cause parts to skip the ramp and collide with the floor or float and there was a lot of code accounting for the fact that the precalc z level wasn't correct until one of the last steps of movement. 

With this, we update the z levels at the same time as the rest of the movement, prior to collisions etc. Rather than keep track of every part going over the ramp, we only track the center part. For other parts we draw a line from the center to the part and look at any ramps that line crosses. There's a check at the start to skip the process if the vehicle isn't touching a ramp. 

Interacting across z levels just needed the part's z level passing through rather than using the player's. The disappearing vehicles is done by making vehicle cache resets always full resets rather than single z level resets, similar to DDA's solution. 

#### Describe alternatives you've considered

I considered using collisions and falling to handle the ramp skipping after it happened, but each system would have been more complicated, only handled one direction and not given me a chance to remove all the ad-hoc z level adjustments.

I've also had vehicles check their z levels when a vehicle cache reset is done. This is mostly a defensive measure, I'm not sure it's needed. It's done there to be sure it's done every time something is loaded but things should already have the correct z levels when they're loaded. That said, for most vehicles all it will involve is checking they don't touch a ramp and it means you can do things like spawn vehicles on a ramp without worry. 

#### Testing

Just manual testing. I gave the steering a good wiggle going back and forth and checked that stuff didn't disappear when I crossed submaps and that it let me remove a part across z levels. It might be that it's still possible to skip the ramp with the center tile, but I was unable to get this to happen when trying. If it does we'll need to use the rotation pivot rather than the center but that's extra complexity that I don't think is needed.

Coolthulhu commented:
Clang-tidy complains about:
```
+ clang-tidy -quiet src/vehicle_move.cpp
5750 warnings generated.
/home/runner/work/Cataclysm-BN/Cataclysm-BN/src/vehicle_move.cpp:1651:28: error: Prefer constructing 'point' from named constant 'point_zero' rather than explicit integer arguments. [cata-use-named-point-constants,-warnings-as-errors]
    auto center = part_at( {0, 0} );
                           ^~~~~~
                           point_zero
```

Coolthulhu commented on tests/vehicle_test.cpp:
```diff
@@ -73,3 +73,29 @@ TEST_CASE( "add_item_to_broken_vehicle_part" )
+
+void check_wreckage( int zlevel );
+
+void check_wreckage( int zlevel )
```
You can mark is as `static`, this way it won't require this extra declaration above.

joveeater commented:
That turned out to be a very good thing. The problem with the rail tests is my code had a huge bug. It was only updating the part positions, never the actual vehicle position. The only visible result I know of was the remember position thing, but I have to imagine there'd be more somewhere. 

I've also done some refactoring to make vehicle code a little easier to use. level_vehicle is gone, all it did was solve the same floating parts problem but it wouldn't kick in until after you fully left the ramp. This means displace_vehicle and advance_precalc_mounts are no longer doing double duty. They now just consistently move the whole vehicle. 

shift_zlevel has been added to vehicles. This will move their global z position and move all of their parts in the opposite direction to compensate. This means that the right sequence for updating a vehicle's position is precalc_mount to do rotation, then adjust_zlevel to spread parts over ramps, then shift_zlevel to make the global z position correct. I've made this available as a single function, refresh_position. 

I've added a new boolean to vehicles that tracks if they're currently attached to the map. This is used by refresh_position to suppress z level stuff when not attached. precalc_mounts has had a behavior change in that it now reset z levels to 0. precalc_mounts was also being used in a few places to initialize facing and pivots. I've replaced this with explicit methods to set the two. This means part z levels will be correct for attached vehicles and 0 for detached ones. 

This does conflict with vehicle holes, I've put a merged version up at https://github.com/joveeater/Cataclysm-BN/tree/merge-check or I can add it to one of the two if it's easier for you.

Coolthulhu commented:
Problems:
* When walking into closed door from ramp up/down, it opens. But closing it doesn't work - you attempt to close the ramp. Examining the vehicle parts up/down ramp is impossible as well.
* Splitting vehicle by destroying it until it has 2+ disconnected parts, when the vehicle is all "ramped up", places the new vehicles in weird spots
* Landing a helicopter mid-ramp makes it behave differently before and after map shift. Before shift, the helicopter will be above ramp. After shift, it will be "ramped" just like cars would. Even before the shift, walking from the "on bridge" to "above ramp" parts will cause you to walk down the ramp.

All of those are quite rare, so not a huge deal, but if you know how to solve them, it would be better to do it before the merge.
If you need this PR to be merged for some other one, and are prepared to do quick fixes if there are bigger problems that I didn't find, I could merge it in current state.

joveeater commented:
1. I've left this. I'm not sure how to approach it well. I don't want to push the actions upwards on ramps and remove the ability to grab items out of the tile etc. 

2. The root cause of this is that if the new vehicle contains the old vehicle's center, the old one ends up disjoint from its own global_pos3(). This causes bad behavior if no part of the vehicle is touching a ramp in which case it's assumed it's flat with its pos3. I will do another pr at some point which shifts the vehicle wrt position (rather than wrt pivot like the current shift_parts), but for now it just treats vehicles with disjoint positions as always requiring proper ramp handling. 

3. Landing and taking off mid ramp are now "unsafe." I had a look at allowing it and landing isn't a problem but taking off means you sometimes need to go two z levels to fully clear the ramp and not end up in the same half-landed state you get in currently. I don't think it's worth the movement code complexity when requiring a flat landing is easily justified in-game. 

4. Another problem I found was that when generating maps sometimes it would deference a null submap when looking for ramps. Theoretically map::generate should have filled in the submap that it's looking for and I have no idea how it ends up null. For now I've just added a null check to the terrain function, but this is something else I'd like to come back to. 

5. Looking at this again after looking at flight it occurred to me that the vertical_move used to move the player when going over a ramp should be vertical_shift. Move subtracts moves. This means no more skipped turn when your player goes over a ramp. 

6. Github's numbered lists are totally broken. When I press return it does this:
7. 8. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7. 7.

Coolthulhu commented on src/vehicle_move.cpp:
```diff
@@ -1410,51 +1416,111 @@ vehicle *vehicle::act_on_map()
+bool vehicle::check_on_ramp( int idir, const tripoint &offset ) const
+{
+    for( auto &prt : get_all_parts() ) {
+        tripoint partPoint = global_pos3() + offset + prt.part().precalc[idir];
```
Nit: naming, use of `auto`

Coolthulhu commented on src/vehicle.h:
```diff
@@ -761,6 +761,14 @@ class vehicle
+
+        inline void detach() {
+            attached = false;
+        }
```
Those are pretty hacky. Looks OK enough for now, but we'll have to handle it differently in the future.

Coolthulhu commented on src/vehicle.cpp:
```diff
@@ -6846,16 +6859,14 @@ void vehicle::calc_mass_center( bool use_precalc ) const
 }
 
-bounding_box vehicle::get_bounding_box()
+bounding_box vehicle::get_bounding_box( )
```
Seems that your IDE may be adding extra spaces in braces.
